### PR TITLE
feat(security): add code-scanning with CodeQL

### LINKED ISSUE #3176: Explore CodeQL or other static code analysis tools
We currently use eslint in our build process, but we might want to explore other tools, such as CodeQL, to catch problematic constructs in our code.

### PR DESCRIPTION
**Pull request by jsjoeio**
This PR adds code-scanning using CodeQL (by GitHub) to help automatically detect common vulnerability and coding errors.

Fixes #3176

### CODE REVIEW

**Review by jawnsy — APPROVED** [review]
Nice! Excellent work Mr. Previte

File: .github/workflows/codeql-analysis.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        uses: github/codeql-action/autobuild@v1
+
+      # ℹ️ Command-line programs to run using the OS shell.
+      # 📚 https://git.io/JvXDl
+
+      # ✏️ If the Autobuild fails above, remove it and uncomment the following three lines
+      #    and modify them (or add more) to build your code if your project
+      #    uses a compiled language
+
+      #- run: |
```
[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:47]: Good call! I will remove!

File: .github/workflows/codeql-analysis.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    # Runs every Monday morning PST
+    - cron: "17 15 * * 1"
+
+jobs:
+  analyze:
+    name: Analyze
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
```
[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:19]: I think you're right. I'll remove!

File: .github/workflows/codeql-analysis.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    # The branches below must be a subset of the branches above
+    branches: [main]
+  schedule:
+    # Runs every Monday morning PST
+    - cron: "17 15 * * 1"
+
+jobs:
+  analyze:
+    name: Analyze
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:16]: Good call! I'll change this

File: .github/workflows/codeql-analysis.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+          languages: javascript
+          # If you wish to specify custom queries, you can do so here or in a config file.
+          # By default, queries listed here will override any specified in a config file.
+          # Prefix the list here with "+" to use these queries and those in the config file.
+          # queries: ./path/to/local/query, your-org/your-repo/queries@main
+
+      # Autobuild attempts to build any compiled languages  (C/C++, C#, or Java).
+      # If this step fails, then you should remove it and run the build manually (see below)
+      - name: Autobuild
+        uses: github/codeql-action/autobuild@v1
```
Comment (jawnsy) [inline .github/workflows/codeql-analysis.yml:38]: If GitHub handles these versions the same way that they do for actions (one tag for v1.1 and one tag for v1) then we might want to add this to dependabot's ignores

Also, it seems that [codeql-action has a v2](https://github.com/github/codeql-action) according to the tags, but idk if that applies to everything or how that all works - I didn't realize that it was possible to have multiple actions per repo!

File: .github/workflows/codeql-analysis.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+      # ✏️ If the Autobuild fails above, remove it and uncomment the following three lines
+      #    and modify them (or add more) to build your code if your project
+      #    uses a compiled language
+
+      #- run: |
+      #   make bootstrap
+      #   make release
+
+      - name: Perform CodeQL Analysis
```
[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:51]: Probably merge and reduce gradually. Seems like a solid approach.

Good idea on ignoring `lib/vscode`. I'll see if that's configurable

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:38]: Ooo good point. I'll see if I can figure that out 😬

**Comment by jsjoeio** [comment]
P.S. - @jawnsy how did you know how to go [here](https://github.com/cdr/code-server/security/code-scanning?page=2&query=ref%3Arefs%2Fpull%2F3229%2Fmerge+tool%3ACodeQL) for the Code scanning alerts?

My intuitive would have told me they show up [here](https://github.com/cdr/code-server/security/code-scanning) but that's not the case 🤷‍♂️

File: .github/workflows/codeql-analysis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jawnsy) [reply .github/workflows/codeql-analysis.yml:51]: @jsjoeio Replying here so it's threaded

> - @jawnsy how did you know to go [here](https://github.com/cdr/code-server/security/code-scanning?page=2&query=ref%3Arefs%2Fpull%2F3229%2Fmerge+tool%3ACodeQL) for the Code scanning alerts?

I found it by clicking looking at the Actions result, then clicking "view all alerts" - super unintuitive lmao

image

[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:51]: I found this in the docs: [Specifying Directories to Scan](https://docs.github.com/en/code-security/secure-coding/configuring-code-scanning#specifying-directories-to-scan)

It's unclear whether this needs to be in a custom codeql config file or if I can put it in the workflow file. I'm going to try adding to the workflow file.

[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:51]: Also opened follow-up issue: https://github.com/cdr/code-server/issues/3243

File: .github/workflows/codeql-analysis.yml (hunk 4 above)
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:31]: Did I do this right?

https://github.com/cdr/code-server/blob/0934accc84aba9b50dc82b7ccee53b7bd9971ec6/.github/dependabot.yml#L12

Comment (jawnsy) [reply .github/workflows/codeql-analysis.yml:31]: I think so? I hope so lol. One can never know with dependabot :joy:

Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:31]: Right? We both know this never works out 100% of the time 😂 well, easily reversible if this is wrong!

File: .github/workflows/codeql-analysis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:33]: And...there's my answer! I'll do it the proper way instead

image

[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/codeql-analysis.yml:32]: It worked! https://github.com/cdr/code-server/security/code-scanning?query=ref%3Arefs%2Fpull%2F3229%2Fmerge+tool%3ACodeQL

**Pull request merged**